.github/workflows/main.yml

```diff
@@ -14,23 +14,70 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        docker-image:  # See https://hub.docker.com/_/python/tags for images
-          - python:2.7.18-alpine3.11
-          - python:3.11.4-alpine3.18
-    container: ${{ matrix.docker-image }}
+        python-version: ['2.7', '3.11']
     steps:
       - name: Checkout code
         uses: actions/checkout@v3
+      - uses: LizardByte/setup-python-action@master
+        with:
+          python-version: ${{matrix.python-version}}
+
+      - uses: actions/cache@v3
+        name: Setup cache for running pre-commit fast
+        with:
+            path: ~/.cache/pre-commit
+            key: pre-commit|${{ env.pythonLocation }}|${{ hashFiles('.pre-commit-config.yaml') }}
+
+      - run: echo "::add-matcher::.github/workflows/python-warning-matcher.json"
+        name: "Setup GitHub for reporting Python warnings as annotations in pull request code review"
+
+      - uses: pre-commit/action@v3.0.0
+        name: Run pre-commit checks (no spaces at end of lines, etc)
+        if: ${{ matrix.python-version != '2.7' }}
+        with:
+            extra_args: --all-files --verbose
+        env:
+          PYTHONDEVMODE: yes
+          SKIP: no-commit-to-branch
 
       - name: Install python 2 dependencies
-        if: ${{ startsWith(matrix.docker-image, 'python:2.7.18') }}
+        if: ${{ matrix.python-version == '2.7' }}
         run: pip install enum
 
-      - name: Install dependencies
-        run: pip install mock pytest
+      - name: Install common dependencies for Python${{matrix.python-version}}
+        run: pip install mock pytest-coverage pytest-mock
+
+      - name: Run Pytest tests for Python${{matrix.python-version}}
+        run: >
+          pytest --cov scripts scripts/ -vv -rA
+          --junitxml=.git/pytest${{matrix.python-version}}.xml
+          --cov-report term-missing
+          --cov-report html:.git/coverage${{matrix.python-version}}.html
+          --cov-report xml:.git/coverage${{matrix.python-version}}.xml
+
+      - name: Upload Python${{matrix.python-version}} coverage reports to Codecov
+        uses: codecov/codecov-action@v3
+        with:
+          directory: .git
+          files: coverage${{matrix.python-version}}.xml
+          env_vars: OS,PYTHON
+          fail_ci_if_error: false
+          flags: python${{matrix.python-version}}
+          name: coverage${{matrix.python-version}}
+          verbose: true
+
+      - name: Add Pytest coverage comment (if write permission is available)
```

Comment:
I believe we have way too many automated comments, the codevoc should be more than enough Can this be removed, please?

Reply:
Done.